src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,438 @@
+        const result: Promise<void> = joinerDAO.addCandidate(partId, candidate);
+
+        // Then it should succeed
+        await expectAsync(result).toBeResolvedTo();
+    });
+    it('should forbid a user to use a fake name in the list of candidates', async() => {
+        // Given a part and joiner, and a verified user (that is not the creator)
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+        await signOut();
+
+        const candidateUser: FireAuth.User = await createConnectedGoogleUser('bar@bar.com', 'candidate');
+
+        // When the user adds themself to the list of candidates with a fake name
+        const result: Promise<void> = joinerDAO.addCandidate(partId, { id: candidateUser.uid, name: 'blibli' });
+
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should forbid a user to add anyone else to the candidates', async() => {
+        // Given a part and joiner, and two verified users (that are not the creator)
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+        await signOut();
+
+        const attackedCandidateUser: FireAuth.User = await createConnectedGoogleUser('bar@bar.com', 'candidate');
+        const attackedCandidate: MinimalUser = { id: attackedCandidateUser.uid, name: 'candidate' };
+        await signOut();
+        await createConnectedGoogleUser('attacker@bar.com', 'attacker');
+
+        // When the user adds someone else to the list of candidates
+        const result: Promise<void> = joinerDAO.addCandidate(partId, attackedCandidate);
+
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should allow a user to remove themself from the candidates', async() => {
+        // Given a part and joiner, and a candidate
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+        await signOut();
+
+        const candidateUser: FireAuth.User = await createConnectedGoogleUser('bar@bar.com', 'candidate');
+        const candidate: MinimalUser = { id: candidateUser.uid, name: 'candidate' };
+        await joinerDAO.addCandidate(partId, candidate);
```

Comment:
c'est très lisible mais je me dit qu'une refactor de ces given genre
given([
    testUser('creator', 'not-connected'),
    testUser('candidate-josé', 'connected'),
    testUser('candidate-jaja', 'not-connected')
]);
Pourrait être sexy non ?

Reply:
bof bof! Car faudrait avoir des créations de parties etc. entre tout ça, donc on se retrouvera à ... faire du code dans une liste, donc autant direct faire le code en question !